### ISSUE TITLE
[Bug]: Browsers sometimes close before starting visual comparison tests

### ISSUE THREAD

**Issue by z0n** [microsoft/playwright#8097]
### Playwright version

1.13.1

### Operating system

Linux

### What browsers are you seeing the problem on?

Chromium, WebKit

### Other information

I'm using the official Playwright Docker image to run the tests.  
The website under test is a Next.js application running using the dev server on the same machine as the tests.

### What happened? / Describe the bug

Sometimes, without any discernible pattern, some of my visual comparison tests immediately fail after 0ms. When I add `retries` to the config, the next try will usually work as expected. It's not always the same tests which fail but it's always visual comparison tests.  
The logs are not very helpful and only show this:
```
immediate._onImmediate: Target page, context or browser has been closed
```
I'm not sure how to debug this as it's very random.  
From what I can tell, this issue only occurs if I run all tests in parallel (Chrome or Webkit; Firefox doesn't seem to have this issue). If I limit it to one test using `test.only()` (-> 6 browsers in parallel but only one test each) *or* one project, e.g. `--project "Chrome 1080p"` (-> 1  browser running all tests), the tests pass every time. 

### Code snippet to reproduce your bug

```shell
// playwright.config.ts
import { devices, PlaywrightTestConfig } from '@playwright/test'

export const DESKTOP_1080P_VIEWPORT = { width: 1920, height: 1080 }

const config: PlaywrightTestConfig = {
    use: {
        baseURL: process.env.BASE_URL ? process.env.BASE_URL : 'http://localhost:3000',
    },
    testDir: './tests/playwright/',
    testMatch: '*.test.ts',
    timeout: 30000,
    preserveOutput: 'always',
    projects: [
        {
            name: 'Chrome 1080p',
            use: {
                browserName: 'chromium',
                launchOptions: { args: ['--disable-dev-shm-usage'] },
                viewport: DESKTOP_1080P_VIEWPORT,
                screenshot: 'only-on-failure',
                video: 'retain-on-failure',
            },
        },
        {
            name: 'Safari 1080p',
            use: {
                browserName: 'webkit',
                viewport: DESKTOP_1080P_VIEWPORT,
                screenshot: 'only-on-failure',
                video: 'retain-on-failure',
            },
        },
        {
            name: 'Firefox 1080p',
            use: {
                browserName: 'firefox',
                viewport: DESKTOP_1080P_VIEWPORT,
                screenshot: 'only-on-failure',
                video: 'retain-on-failure',
            },
        },
        // Test against mobile viewports.
        {
            name: 'Chrome Pixel 5',
            use: {
                ...devices['Pixel 5'],
                screenshot: 'only-on-failure',
                video: 'retain-on-failure',
            },
        },
        {
            name: 'Safari iPhone 12',
            use: {
                ...devices['iPhone 12'],
                screenshot: 'only-on-failure',
                video: 'retain-on-failure',
            },
        },
        {
            name: 'Safari iPhone SE',
            use: {
                ...devices['iPhone SE'],
                screenshot: 'only-on-failure',
                video: 'retain-on-failure',
            },
        },
    ],
}
export default config

// example tests which are failing randomly
import { expect, test } from '@playwright/test'
import { HomePage } from './home-page.page'

test.describe('Home page', () => {
    test.beforeEach(async ({ page }) => {
        const homePage = new HomePage(page)
        await homePage.goto()
    })

    test('should look as expected', async ({ page }) => {
        test.skip(process.platform !== 'linux', 'Visual comparisons are only done on Linux/Docker')
        expect(await page.screenshot({ fullPage: true })).toMatchSnapshot('home-page.png', { threshold: 0.1 })
    })

    test('focus on search input should look as expected', async ({ page }) => {
        test.skip(process.platform !== 'linux', 'Visual comparisons are only done on Linux/Docker')
        const homePage = new HomePage(page)
        await homePage.focusOnSearchInput()
        expect(await page.screenshot({ fullPage: true })).toMatchSnapshot('home-page-search-input-focus.png', {
            threshold: 0.05,
        })
    })
})

// home-page.page.ts
import { Page } from '@playwright/test'
import { goto } from '../../utils/goto'

export class HomePage {
    readonly page: Page
    readonly searchInputSelector: string

    constructor(page: Page) {
        this.page = page
        this.searchInputSelector = 'input[name="search"]'
    }

    async waitForPageReady() {
        return this.page.waitForSelector(this.searchInputSelector)
    }

    async goto() {
        await goto(this.page, '/')
        await this.waitForPageReady()
    }

    async focusOnSearchInput() {
        await this.page.focus(this.searchInputSelector)
    }

// goto.ts
import { Page } from '@playwright/test'
import { disableAnimations } from './disable-animations'

export const goto = async (
    page: Page,
    url: string,
    options: {
        referer?: string
        timeout?: number
        waitUntil?: 'load' | 'domcontentloaded' | 'networkidle'
    } = { waitUntil: 'networkidle' },
) => {
    await page.route(/\/webpack-hmr/, route => route.abort()) // abort the webpack request to allow using 'networkidle' when using the dev server
    await page.goto(url, options)
    await disableAnimations(page)
    await this.page.waitForSelector('input[name="search"])
}

// disable-animations.ts
import { Page } from '@playwright/test'

export const disableAnimations = async (page: Page): Promise<void> => {
    await page.addStyleTag({
        content: `*,
        *::before,
        *::after {
            -moz-animation: none !important;
            -moz-transition: none !important;
            animation: none !important;
            caret-color: transparent !important;
            transition: none !important;
        }`,
    })
}
```

### Relevant log output

```shell
2) [Chrome Pixel 5] › pages/home-page/home-page.test.ts:11:5 › Home page should look as expected =

    immediate._onImmediate: Target page, context or browser has been closed
```

**Comment by aslushnikov** [comment]
@z0n could you please share how do start docker image? It might be that shm memory is not enough.

**Comment by z0n** [comment]
@aslushnikov Sure, here's the command I'm using in package.json (the tests are executed in the "nextjs" workspace:
```
docker-compose -f ./tests/playwright/docker-compose.yml run --rm playwright yarn workspace nextjs playwright test --output reports/playwright
```

Here's my docker-compose.yml file:
```
version: "3.9"

services:
    playwright:
        image: mcr.microsoft.com/playwright:v1.13.1
        environment:
            BASE_URL: ${BASE_URL:-http://host.docker.internal:3000}
        extra_hosts:
            - host.docker.internal:host-gateway
        stdin_open: true
        tty: true
        volumes:
            - ../../../../:/opt/nextjs
        working_dir: /opt/nextjs
```

I thought /dev/shm was not needed as I set `--disable-dev-shm-usage` for Chrome though? I can try with it enabled, I'll just have to figure out how this will work in macOS (I'm running the container in Docker Desktop for macOS, forgot to mention that).

**Comment by z0n** [comment]
One more thing: My Docker resource configuration:
<img width="907" alt="image" src="https://user-images.githubusercontent.com/7430908/128870822-bd12849c-0d09-46fd-8569-4b59626b347f.png">

**Comment by pavelfeldman** [comment]
Sorry about the delay. I don't think we can help without the repro. Did you have any progress on debugging this? I would enable all kinds of logging `DEBUG=pw:api,pw:browser*` and see if something is unusual there.

**Comment by pavelfeldman** [comment]
Sadly, we can't act on this without extra info. Please file a new issue and link this one if you get back to it!
